# Pandas Training Book

### Why I made this notebook?

This Notebook is here to train the Pandas Library. 
It contains quizzles for all fundamental Pandas attributes, methods and more. 
The goal is to **master the mechanics through repetition**. 
The work here is based on the book "Python Data Science Handbook" written by Jack VanderPlas, which I really enjoyed reading.

### Requirements

You should know the theory behind pandas and have an understanding why the following tasks work.
This book is just about knowing how to use the code.

### Let's get started

#### Task 1 Installation

1.1 Import Pandas

In [1]:
import pandas as pd

In [2]:
#enter solution


1.2 Show Pandas version

In [3]:
pd.__version__

'1.0.3'

In [4]:
#enter solution


1.3 Import the other library, that pandas works with

In [5]:
import numpy as np

In [6]:
#enter solution


#### Task 2 Introducing Pandas Objects

2.1 Create a pandas series with the values 1, 2 , 3 , 4 and save it in the variable **data**

In [7]:
data = pd.Series([1, 2, 3, 4])
data

0    1
1    2
2    3
3    4
dtype: int64

In [8]:
#enter solution


2.2 Get the values from the **data** object

In [9]:
data.values

array([1, 2, 3, 4], dtype=int64)

In [10]:
#enter solution


2.3 Show the index-range from the **data** object

In [11]:
data.index

RangeIndex(start=0, stop=4, step=1)

In [12]:
#enter solution


2.4 Get the second item from the **data** object

In [13]:
data[1]

2

In [14]:
#enter solution


2.5 Get the values from index 1 to 2

In [15]:
data[1:3]

1    2
2    3
dtype: int64

In [16]:
#enter solution


2.6 Create a pandas series with the values 1, 2, 3, 4 and indices a, b, c, d and save it in the varibale **data**

In [17]:
data = pd.Series([1,2,3,4], index=["a", "b", "c", "d"])
data

a    1
b    2
c    3
d    4
dtype: int64

In [18]:
#enter solution


2.7 Get the item for the index "b"

In [19]:
data["b"]

2

In [20]:
#enter solution


2.8 Create a dictionary with the keys "Person1", "Person2", "Person3", "Person4" and the values, 12 ,13 ,15 ,11
save it as **person_dict**. Create a pandas series out of this dictionary and save it as **person_pd**.

In [21]:
person_dict = {"Person1": 12, "Person2": 13, "Person3": 15, "Person4": 11}
print(person_dict)
person_pd = pd.Series(person_dict)
person_pd

{'Person1': 12, 'Person2': 13, 'Person3': 15, 'Person4': 11}


Person1    12
Person2    13
Person3    15
Person4    11
dtype: int64

In [22]:
#enter solution


2.9 Get the values for Person3

In [23]:
person_pd["Person3"]

15

In [24]:
#enter solution


2.10 Get the values from Person2 to Person4

In [25]:
person_pd["Person2":"Person4"]

Person2    13
Person3    15
Person4    11
dtype: int64

In [26]:
#enter solution


2.11 Get a pandas series from the following dictionary with only index 3 and 2, save as variable named data

In [27]:
test_dict = {2: "a", 3: "c", 1: "a"}

data = pd.Series(test_dict, index= [3,2])
data

3    c
2    a
dtype: object

In [28]:
test_dict = {2: "a", 3: "c", 1: "a"}

#enter solution


#### Task 3 Pandas DataFrame Object

3.1 Create two dictionaries named person_number_dict and person_wage_dict with the keys "Person1", "Person2", "Person3", "Person4". 
The values for person_number_dict: 123 , 124 , 125 , 126
The values for person_wage_dict: 10000, 12000 , 9000, 5000

In [29]:
person_number_dict = {"Person1": 123, "Person2": 124, "Person3": 125, "Person4": 126}
person_wage_dict = {"Person1": 10000, "Person2": 12000, "Person3": 9000, "Person4": 5000}

In [30]:
#enter solution


3.2 Create two pandas series with the dictionaries before. Named person_number_pd and person_wage_pd.

In [31]:
person_number_pd = pd.Series(person_number_dict)
person_wage_pd = pd.Series(person_wage_dict)

In [32]:
#enter solution


3.3 Create a data frame with the index set as "Person1", "Person2" ... etc. and the column named as "per. number" and "wage". Fill in the pandas series from before. Save it in the variable **data**

In [33]:
data = pd.DataFrame({"per. number": person_number_pd, "wage": person_wage_pd})
print(data)

         per. number   wage
Person1          123  10000
Person2          124  12000
Person3          125   9000
Person4          126   5000


In [34]:
#enter solution


3.4 Get the indices from the **data** set

In [35]:
data.index

Index(['Person1', 'Person2', 'Person3', 'Person4'], dtype='object')

In [36]:
#enter solution


3.5 Get the column names from the **data** set

In [37]:
data.columns

Index(['per. number', 'wage'], dtype='object')

In [38]:
#enter solution


3.6 Get the column for "wage" from the **data** set

In [39]:
data["wage"]

Person1    10000
Person2    12000
Person3     9000
Person4     5000
Name: wage, dtype: int64

In [40]:
#enter solution


#### Task 4 Create Panda DataFrames

4.1 Create a Panda DataFrame from **person_wage_pd**

In [41]:
data = pd.DataFrame(person_wage_pd, columns= ["wage"])
data

wage
Person1  10000
Person2  12000
Person3   9000
Person4   5000

In [42]:
#enter solution


4.2 Create a Panda DataFrame from the given dictionary

In [43]:
x_dict = [{"a":i ,"b":i*2 , "c": i**3} for i in range(3)]

pd.DataFrame(x_dict)

a  b  c
0  0  0  0
1  1  2  1
2  2  4  8

In [44]:
x_dict = [{"a":i ,"b":i*2 , "c": i**3} for i in range(3)]

#enter solution


4.3 Create a Panda DataFrame from the two following dictionaries

In [45]:
x_dict = {"a": 1, "b": 2}
y_dict = {"b": 4, "c": 3}

x_pd = pd.DataFrame([x_dict, y_dict])
x_pd

a  b    c
0  1.0  2  NaN
1  NaN  4  3.0

In [46]:
x_dict = {"a": 1, "b": 2}
y_dict = {"b": 4, "c": 3}

#enter solution


4.4 Create a Panda DataFrame from the following Numpy array and lists

In [47]:
np.random.seed(42)
np_data = np.random.rand(3,2)
cols = ["foo", "bar"]
ind = ["a", "b", "c"]

pd.DataFrame(np_data, columns=cols, index= ind)

foo       bar
a  0.374540  0.950714
b  0.731994  0.598658
c  0.156019  0.155995

In [48]:
np.random.seed(42)
np_data = np.random.rand(3,2)
cols = ["foo", "bar"]
ind = ["a", "b", "c"]

#enter solution


4.5 Get the intersection of the follwing two indices

In [49]:
indA = pd.Index([1,3,5,7,9])
indB = pd.Index([2,3,5,7,11])

indA & indB

Int64Index([3, 5, 7], dtype='int64')

In [50]:
indA = pd.Index([1,3,5,7,9])
indB = pd.Index([2,3,5,7,11])

#enter solution


4.6 Get the union of **indA** and **indB**

In [51]:
indA | indB

Int64Index([1, 2, 3, 5, 7, 9, 11], dtype='int64')

In [52]:
#enter solution


4.7 Get the symmetric difference of **indA** and **indB**

In [53]:
indA ^ indB

Int64Index([1, 2, 9, 11], dtype='int64')

In [54]:
#enter solution


#### Task 5 Data Indexing and Selection 

5.1 Add 1.25 as index "e" to the Series below

In [55]:
data = pd.Series([0.25, 0.5, 0.75, 1.0], index= ["a", "b", "c", "d"])
data["e"] = 1.25
data

a    0.25
b    0.50
c    0.75
d    1.00
e    1.25
dtype: float64

In [56]:
data = pd.Series([0.25, 0.5, 0.75, 1.0], index= ["a", "b", "c", "d"])

#enter solution


5.2 Get all data between 0.3 and 0.8 from **data**

In [57]:
data[(data > 0.3) & (data < 0.8)]

b    0.50
c    0.75
dtype: float64

In [58]:
#enter solution


5.3 Get the explicit index 1 from the following data set

In [59]:
data = pd.DataFrame(["a", "b", "c", "d", "e"], index= [1,2,3,4,5],columns= ["First"])
print(data)
data.loc[1]

  First
1     a
2     b
3     c
4     d
5     e


First    a
Name: 1, dtype: object

In [60]:
data = pd.DataFrame(["a", "b", "c", "d", "e"], index= [1,2,3,4,5],columns= ["First"])

#enter solution


5.4 Get the implicit index 1 from **data**

In [61]:
data.iloc[1]

First    b
Name: 2, dtype: object

In [62]:
#enter solution


5.5 Add a column "Yearly wage" with the wages x12 for the following dataset.

In [63]:
person_wage_dict = {"Person1": 10000, "Person2": 12000, "Person3": 7000, "Person4":5000}
person_number_dict = {"Person1": 123, "Person2": 124, "Person3": 125, "Person4": 126}
person_wage_pd = pd.Series(person_wage_dict)
person_number_pd = pd.Series(person_number_dict)
data = pd.DataFrame({"Person Number": person_number_pd,"Person wage": person_wage_pd})
print(data)

data["Yearly wage"] = data["Person wage"] * 12
data


         Person Number  Person wage
Person1            123        10000
Person2            124        12000
Person3            125         7000
Person4            126         5000


Person Number  Person wage  Yearly wage
Person1            123        10000       120000
Person2            124        12000       144000
Person3            125         7000        84000
Person4            126         5000        60000

In [64]:
person_wage_dict = {"Person1": 10000, "Person2": 12000, "Person3": 7000, "Person4":5000}
person_number_dict = {"Person1": 123, "Person2": 124, "Person3": 125, "Person4": 126}
person_wage_pd = pd.Series(person_wage_dict)
person_number_pd = pd.Series(person_number_dict)
data = pd.DataFrame({"Person Number": person_number_pd,"Person wage": person_wage_pd})

#enter solution


5.6 Get the values from **data** as a NumPy array

In [65]:
data.values

array([[  123, 10000],
       [  124, 12000],
       [  125,  7000],
       [  126,  5000]], dtype=int64)

In [66]:
#enter solution


5.7 Transpose **data**

In [67]:
data.T

Person1  Person2  Person3  Person4
Person Number      123      124      125      126
Person wage      10000    12000     7000     5000

In [68]:
#enter solution


5.8 Get the second row from **data** as a NumPy array

In [69]:
data.values[1]

array([  124, 12000], dtype=int64)

In [70]:
#enter solution


5.9 Get the column for ***Person wage*** from **data**

In [71]:
data["Person wage"]

Person1    10000
Person2    12000
Person3     7000
Person4     5000
Name: Person wage, dtype: int64

In [72]:
#enter solution


5.10 Get the implicit indices for the first three rows and first two columns from **data**

In [73]:
data.iloc[:3, :2]

Person Number  Person wage
Person1            123        10000
Person2            124        12000
Person3            125         7000

In [74]:
#enter solution


5.11 Get the data for Person1 to Person3 and the columns Person number and Person wage from **data**

In [75]:
data.loc[:"Person3", : "Person wage"]

Person Number  Person wage
Person1            123        10000
Person2            124        12000
Person3            125         7000

In [76]:
#enter solution


#### Task 6 Operations in Pandas

6.1 Add the following Series together and fill the missing values in A or B with 0

In [77]:
A = pd.Series([2,4,6], index=[0,1,2])
B = pd.Series([1,3,5], index=[1,2,3])

A.add(B, fill_value= 0)

0    2.0
1    5.0
2    9.0
3    5.0
dtype: float64

In [78]:
A = pd.Series([2,4,6], index=[0,1,2])
B = pd.Series([1,3,5], index=[1,2,3])

#enter solution


6.2 Add the following Series together and replace missing values in **A** with the mean of **A.stack()**

In [79]:
A = pd.DataFrame(np.random.randint(0,20, (2,2)),columns=list("AB"))
B = pd.DataFrame(np.random.randint(0,10, (3,3)),columns=list("BAC"))
print(A)
print(B)
fill = A.stack().mean()
print(A.add(B, fill_value=fill))

    A   B
0  10  10
1   3   7
   B  A  C
0  7  2  5
1  4  1  7
2  5  1  4
      A     B     C
0  12.0  17.0  12.5
1   4.0  11.0  14.5
2   8.5  12.5  11.5


In [80]:
A = pd.DataFrame(np.random.randint(0,20, (2,2)),columns=list("AB"))
B = pd.DataFrame(np.random.randint(0,10, (3,3)),columns=list("BAC"))
print(A)
print(B)

#enter solution


    A   B
0   0  11
1  11  16
   B  A  C
0  9  2  6
1  3  8  2
2  4  2  6


6.3 Substract **B** from **A** (use Pandas operator and Pandas method)

In [81]:
A - B
A.sub(B)

A     B   C
0 -2.0   2.0 NaN
1  3.0  13.0 NaN
2  NaN   NaN NaN

In [82]:
#enter solution


6.4 Multiply **A** with **B** (use Pandas operator and Pandas method)

In [83]:
A * B
A.mul(B)

A     B   C
0   0.0  99.0 NaN
1  88.0  48.0 NaN
2   NaN   NaN NaN

In [84]:
#enter solution


6.5 Divide **A** by **B** (use Pandas operator and Pandas method)

In [85]:
A / B
A.div(B)

A         B   C
0  0.000  1.222222 NaN
1  1.375  5.333333 NaN
2    NaN       NaN NaN

In [86]:
#enter solution


6.6 Get the modulus of **A** divided by **B** (use Pandas operator and Pandas method)

In [87]:
A % B
A.mod(B)

A    B   C
0  0.0  2.0 NaN
1  3.0  1.0 NaN
2  NaN  NaN NaN

In [88]:
#enter solution


6.7 Get **A** to the power of **B** (use Pandas operator and Pandas method)

In [89]:
A ** B
A.pow(B)

A             B   C
0          0.0  2.357948e+09 NaN
1  214358881.0  4.096000e+03 NaN
2          NaN           NaN NaN

In [90]:
#enter solution


#### Task 7 Handling missing data

7.1 Get the sum of the following array

In [91]:
data = np.array([1, np.nan, 3, 4])

np.nansum(data)

8.0

In [92]:
data = np.array([1, np.nan, 3, 4])

#enter solution


7.2 Get the minimum from **data**

In [93]:
np.nanmin(data)

1.0

In [94]:
#enter solution


7.3 Get the maximum from **data**

In [95]:
np.nanmax(data)

4.0

In [96]:
#enter solution


7.4 Check which elements are Null values in the following Series

In [97]:
data = pd.Series([1, np.nan, "hello", None])

data.isnull()

0    False
1     True
2    False
3     True
dtype: bool

In [98]:
data = pd.Series([1, np.nan, "hello", None])

#enter solution


7.5 Get the elements in a new Series without Null values

In [99]:
data[data.isnull() == False]
#or
data[data.notnull()]
#or
data.dropna()

0        1
2    hello
dtype: object

In [100]:
#enter solution


7.6 Drop the elements that are Null values from the following DataFrame

In [101]:
data = pd.DataFrame([[1, np.nan, 2 ],
                     [2, 3 ,5 ],
                     [np.nan, 4, 6 ]])
print(data)
data.dropna()

     0    1  2
0  1.0  NaN  2
1  2.0  3.0  5
2  NaN  4.0  6


0    1  2
1  2.0  3.0  5

In [102]:
data = pd.DataFrame([[1, np.nan, 2 ],
                     [2, 3 ,5 ],
                     [np.nan, 4, 6 ]])
print(data)

#enter solution


     0    1  2
0  1.0  NaN  2
1  2.0  3.0  5
2  NaN  4.0  6


7.7 Drop the columns with Null value elements from **data**

In [103]:
data.dropna(axis= "columns")

2
0  2
1  5
2  6

In [104]:
#enter solution


7.8 Add a column with Null value to **data**

In [105]:
data[3] = np.nan
data

0    1  2   3
0  1.0  NaN  2 NaN
1  2.0  3.0  5 NaN
2  NaN  4.0  6 NaN

In [106]:
#enter solution


7.9 Drop only the columns where all elements are Null value from **data**

In [107]:
data.dropna(axis= "columns", how= "all")

0    1  2
0  1.0  NaN  2
1  2.0  3.0  5
2  NaN  4.0  6

In [108]:
#enter solution


7.10 Fill the following Null values with zeros

In [109]:
data = pd.Series([1, np.nan, 2, None, 3], index= list("abcde"))
print(data)
data.fillna(0)

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64


a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

In [110]:
data = pd.Series([1, np.nan, 2, None, 3], index= list("abcde"))
print(data)

#enter solution


a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64


7.11 Forward-fill the missing data with the previous value

In [111]:
data.fillna(method= "ffill")

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

In [112]:
#enter solution


7.12 Back-fill the missing data with the next value backwards

In [113]:
data.fillna(method= "bfill")

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

In [114]:
#enter solution


#### Task 8 Hierachical Indexing

8.1  Set a multi-index with the following list of tuples

In [115]:
index = [("California", 2000), ("California", 2010), ("New York", 2000), 
         ("New York", 2010), ("Texas", 2000), ("Texas", 2010)]

index = pd.MultiIndex.from_tuples(index)
print(index)

MultiIndex([('California', 2000),
            ('California', 2010),
            (  'New York', 2000),
            (  'New York', 2010),
            (     'Texas', 2000),
            (     'Texas', 2010)],
           )


In [116]:
index = [("California", 2000), ("California", 2010), ("New York", 2000), 
         ("New York", 2010), ("Texas", 2000), ("Texas", 2010)]

#enter solution


8.2 Create a Series with the following populations and multi-index **index**, save as variable named pop

In [117]:
populations = [33871648, 37253956, 18976457, 19378102, 20851820, 25145561]

pop = pd.Series(populations, index = index)
pop

(California, 2000)    33871648
(California, 2010)    37253956
(New York, 2000)      18976457
(New York, 2010)      19378102
(Texas, 2000)         20851820
(Texas, 2010)         25145561
dtype: int64

In [118]:
populations = [33871648, 37253956, 18976457, 19378102, 20851820, 25145561]

#enter solution


8.3 Get the data for all cities in the year 2010 from pop

In [119]:
pop[:, 2010]

ValueError: Can only tuple-index with a MultiIndex

In [ ]:
#enter solution


8.4 Convert the multi-index **pop** to a simple DataFrame

In [ ]:
pop_df = pop.unstack()
pop_df

In [ ]:
#enter solution


8.5 Convert the DataFrame back to a multi-index DataFrame

In [ ]:
pop_df = pop_df.stack()
pop_df

In [ ]:
#enter solution


8.6 Add the following list as a new column to the **pop_df** and name it "under18", name the first column "total"

In [ ]:
under18 = [9267089, 9284094, 4687374, 4318033, 5906301, 6879014]

pop_df = pd.DataFrame({"total": pop, "under18": under18})
pop_df

In [ ]:
under18 = [9267089, 9284094, 4687374, 4318033, 5906301, 6879014]

#enter solution


8.7 Add a third column that shows the fraction in percent numbers of under18 from the total and name it "fu18"

In [ ]:
pop_df["fu18"] = (pop_df["under18"] / pop_df["total"]) *100 
pop_df

In [ ]:
#enter solution


8.8 Create a multi-index from arrays using:
   
    [["a", "a", "b", "b"], [1,2,1,2]]

In [ ]:
[["a", "a", "b", "b"], [1,2,1,2]]

pd.MultiIndex.from_arrays([["a", "a", "b", "b"], [1,2,1,2]])

In [ ]:
#enter solution


8.9 Create the same multi-index from tuples

In [ ]:
pd.MultiIndex.from_tuples([("a",1),("a", 2),("b", 1), ("b",2)])

In [ ]:
#enter solution


8.10 Create the same multi-index from a Cartesian product

In [ ]:
pd.MultiIndex.from_product([["a", "b"], [1,2]])

In [ ]:
#enter solution


8.11 Create the same multi-index from using levels and codes

In [ ]:
pd.MultiIndex(levels=[["a","b"],[1,2]], codes= [[0,0,1,1], [0,1,0,1]])

# In the book the key-word says "labels" instead "codes". If you read the docstring by putting the command 
# below you can see, that it got updated.
#pd.MultiIndex?

In [ ]:
#enter solution


8.12 Set the names attribute for the two indices from **pop_df** "state" and "year"

In [ ]:
pop_df.index.names= ["state", "year"]
pop_df

In [ ]:
#enter solution


8.13 Set the multi-index for index and columns with the following code provided

In [127]:
index = pd.MultiIndex.from_product([[2013, 2014], [1,2]], names= ["year", "visit"])
columns = pd.MultiIndex.from_product([["Bob", "Guido", "Sue"], ["HR", "Temp"]], names= ["subject", "type"])
data = np.round(np.random.randn(4,6), 1)
data[:,::2] *= 10
data += 37

df = pd.DataFrame(data, index= index, columns= columns)
print(df)

subject      Bob       Guido         Sue      
type          HR  Temp    HR  Temp    HR  Temp
year visit                                    
2013 1      43.0  37.9  46.0  37.8  35.0  36.4
     2      22.0  38.1  35.0  36.6  49.0  36.1
2014 1      45.0  37.3  27.0  36.9  47.0  37.8
     2      52.0  37.3  71.0  36.1  33.0  36.9


In [121]:
index = pd.MultiIndex.from_product([[2013, 2014], [1,2]], names= ["year", "visit"])
columns = pd.MultiIndex.from_product([["Bob", "Guido", "Sue"], ["HR", "Temp"]], names= ["subject", "type"])
data = np.round(np.random.randn(4,6), 1)
data[:,::2] *= 10
data += 37

#enter solution


8.14 Get the full data for "Guido"

In [128]:
df["Guido"]

type          HR  Temp
year visit            
2013 1      46.0  37.8
     2      35.0  36.6
2014 1      27.0  36.9
     2      71.0  36.1

In [ ]:
#enter solution


8.15 Get the data for "California" in the year 2000 for **pop**

In [ ]:
pop["California", 2000]

In [ ]:
#enter solution


8.16 Get the full data for "California"

In [ ]:
pop["California"]

In [ ]:
#enter solution


8.17 Get the data for "California" to "New York"

In [ ]:
pop.loc["California": "New York"]

In [ ]:
#enter solution


8.18 Get all data for the year 2000

In [ ]:
pop[:, 2000]

In [ ]:
#enter solution


8.19 Get all data that is over 22'000'000

In [ ]:
pop[pop > 22000000]

In [ ]:
#enter solution


8.20 Get the data for "California" and "Texas"

In [ ]:
#enter solution


In [ ]:
pop[["California", "Texas"]]

8.21 Get the data for subject "Bob" and type "HR" from **df**

In [129]:
df["Bob", "HR"]

year  visit
2013  1        43.0
      2        22.0
2014  1        45.0
      2        52.0
Name: (Bob, HR), dtype: float64

In [ ]:
#enter solution


8.22 Get the data for the first two rows and first two columns 

In [ ]:
df.iloc[:2, :2]

In [ ]:
#enter solution


8.23 Get the data for rows: all "years" and only "visit" 2 and columns: all "subjects" and only type "HR"

In [ ]:
idx = pd.IndexSlice
df.loc[idx[:,2], idx[:,"HR"]]

In [ ]:
#enter solution


8.24 Sort the index alphabetically

In [ ]:
index = pd.MultiIndex.from_product([["a", "c", "b"], [1,2]])
data = pd.Series(np.random.rand(6), index= index)
data.index.names=["char", "int"]
print(data)
data = data.sort_index()
data

In [ ]:
index = pd.MultiIndex.from_product([["a", "c", "b"], [1,2]])
data = pd.Series(np.random.rand(6), index= index)
data.index.names=["char", "int"]
print(data)

#enter solution


8.25 Unstack **pop** at the first index level ("state")

In [ ]:
print(pop)
pop.unstack(level=0)

In [ ]:
#enter solution


8.26 Unstack **pop** at the second index level ("year")

In [ ]:
print(pop)
pop.unstack(level=1)

In [ ]:
#enter solution


8.27 Reset the index as columns in **pop** and save it in **pop_flat**, name the column with population data "population"

In [ ]:
pop_flat = pop.reset_index(name= "population")
pop_flat

In [ ]:
#enter solution


8.28 Set the indices back to "state" and "year" in **pop_flat**

In [ ]:
pop_flat.set_index(["state", "year"])

In [ ]:
#enter solution


8.29 Get the mean for the visits in each year for every subject and type from **df**.

In [ ]:
print(df)
df_mean = df.mean(level="year")
df_mean

In [ ]:
#enter solution


8.30 Get the mean for all type "HR" and "Temp" in each year

In [ ]:
df_mean.mean(axis=1, level="type")

In [ ]:
#enter solution


#### Task 9 Concatenation of NumPy Arrays and Pandas

9.1 Concatenate the following lists horizontally to a NumPy array

In [ ]:
x= [1,2,3]
y= [4,5,6]
z= [7,8,9]
np.concatenate([x,y,z])

In [ ]:
x= [1,2,3]
y= [4,5,6]
z= [7,8,9]

#enter solution


9.2 Concatenate the following list of lists with itself horizontally to a NumPy array

In [ ]:
x= [[1,2],
   [3,4]]
np.concatenate([x,x], axis=1)

In [ ]:
x= [[1,2],
   [3,4]]

#enter solution


9.3 Concatenate the two Panda Series

In [ ]:
ser1 = pd.Series(["A", "B", "C"], index=[1,2,3])
ser2 = pd.Series(["D", "E", "F"], index=[4,5,6])
pd.concat([ser1, ser2])

In [ ]:
ser1 = pd.Series(["A", "B", "C"], index=[1,2,3])
ser2 = pd.Series(["D", "E", "F"], index=[4,5,6])

#enter solution


9.4 Change the index so ser2 gets added horizontally to ser1

In [ ]:
ser1 = pd.Series(["A", "B", "C"], index=[1,2,3])
ser2 = pd.Series(["D", "E", "F"], index=[1,2,3])
pd.concat([ser1,ser2], axis=1)

In [ ]:
#enter solution

ser1 = pd.Series(["A", "B", "C"], index=[1,2,3])
ser2 = pd.Series(["D", "E", "F"], index=[4,5,6])


9.5 Concatenate the two following DataFrames

In [ ]:
#function to create a DataFrame quickly
def make_df(cols,ind):
    """Quickly make a DataFrame"""
    data= {c: [str(c) + str(i) for i in ind]
          for c in cols}
    return pd.DataFrame(data,ind)

df1 = make_df("AB", [1,2])
df2 = make_df("AB", [3,4])
print(df1), print(df2) , print(pd.concat([df1,df2]))

In [ ]:
#function to create a DataFrame quickly
def make_df(cols,ind):
    """Quickly make a DataFrame"""
    data= {c: [str(c) + str(i) for i in ind]
          for c in cols}
    return pd.DataFrame(data,ind)

df1 = make_df("AB", [1,2])
df2 = make_df("AB", [3,4])

#enter solution


9.6 Concatenate the two following DataFrames in the same index

In [ ]:
df3 = make_df("AB", [0,1])
df4 = make_df("CD", [0,1])
print(df3), print(df4), print(pd.concat([df3,df4], axis= "columns"))

In [ ]:
df3 = make_df("AB", [0,1])
df4 = make_df("CD", [0,1])

#enter solution


9.7 Set the index of the following DataFrame **y** to the same index as **x** and concatenate the two together.

In [ ]:
x = make_df("AB", [0,1])
y = make_df("AB", [2,3])

y.index = x.index
print(x), print(y), print(pd.concat([x,y]))

In [ ]:
x = make_df("AB", [0,1])
y = make_df("AB", [2,3])

#enter solution


9.8 Concatenate the previous two df's again, but with the right index

In [ ]:
y.index = x.index
pd.concat([x,y], ignore_index=True)

In [ ]:
#enter solution


9.9 Concatenate the previous two df's again, but set keys x and y to the index series

In [ ]:
pd.concat([x,y], keys=["x", "y"])

In [ ]:
#enter solution


9.10 Concatenate the two following df's

In [ ]:
df5 = make_df("ABC", [1,2])
df6 = make_df("BCD", [3,4])

print(df5), print(df6), print(pd.concat([df5,df6]))

In [ ]:
df5 = make_df("ABC", [1,2])
df6 = make_df("BCD", [3,4])

#enter solution


9.11 Concatenate the columns of the previous two df's that they have in common

In [ ]:
pd.concat([df5,df6], join="inner")

In [ ]:
#enter solution


9.12 Append **df2** to **df1**

In [ ]:
df1.append(df2)

In [ ]:
#enter solution


#### Task 10 Combining Datasets: Merge and Join

10.1  Merge the two following df's and save as df3

In [ ]:
df1 = pd.DataFrame({"employee": ["Bob", "Jake", "Lisa", "Sue"], 
                    "group": ["Accounting", "Engineering", "Engineering", "HR"]})
df2 = pd.DataFrame({"employee": ["Lisa", "Bob", "Jake", "Sue"], "hire_date": [2004,2008,2012,2014]})
print(df1),print(df2)
df3 = df1.merge(df2)
print(df3)

In [ ]:
df1 = pd.DataFrame({"employee": ["Bob", "Jake", "Lisa", "Sue"], 
                    "group": ["Accounting", "Engineering", "Engineering", "HR"]})
df2 = pd.DataFrame({"employee": ["Lisa", "Bob", "Jake", "Sue"], "hire_date": [2004,2008,2012,2014]})

#enter solution


10.2 Merge **df3** with the following df

In [ ]:
df4 = pd.DataFrame({"group": ["Accounting", "Engineering", "HR"], "supervisor": ["Carly", "Guido", "Steve"]})
print(df3), print(df4)
df3.merge(df4)

In [ ]:
df4 = pd.DataFrame({"group": ["Accounting", "Engineering", "HR"], "supervisor": ["Carly", "Guido", "Steve"]})

#enter solution


10.3 Merge **df1** with the following df

In [ ]:
df5 = pd.DataFrame({"group": ["Accounting", "Accounting", "Engineering", "Engineering", "HR", "HR"], 
                    "skills": ["math", "spreadsheet", "coding", "linux", "spreadsheet", "organization"]})
df1.merge(df5)

In [ ]:
df5 = pd.DataFrame({"group": ["Accounting", "Accounting", "Engineering", "Engineering", "HR", "HR"], 
                    "skills": ["math", "spreadsheet", "coding", "linux", "spreadsheet", "organization"]})

#enter solution


10.4 Merge **df1** and **df2** by the column "employee"

In [ ]:
print(df1), print(df2)
print(pd.merge(df1,df2, on="employee"))

In [ ]:
#enter solution


10.5 Merge df1 with the following df considering different labels for the same column

In [ ]:
df3 = pd.DataFrame({"name": ["Bob", "Jake", "Lisa", "Sue"], 
                    "salary": [70000, 80000, 120000, 90000]})
print(df1), print(df3), print(pd.merge(df1,df3, left_on="employee", right_on="name"))

In [ ]:
df3 = pd.DataFrame({"name": ["Bob", "Jake", "Lisa", "Sue"], 
                    "salary": [70000, 80000, 120000, 90000]})

#enter solution


10.6 Merge df1 with df3 considering different labels for the same column and drop the column "name"

In [ ]:
pd.merge(df1,df3, left_on= "employee", right_on= "name").drop("name", axis=1)

In [ ]:
#enter solution


10.7 Set the indices from **df1** and **df2** to "employee" and save it as **df1a** and **df2a** 

In [ ]:
df1a = df1.set_index("employee")
df2a = df2.set_index("employee")

print(df1a), print(df2a)

In [ ]:
#enter solution


10.8 Merge **df1a** and **df2a** by index

In [ ]:
pd.merge(df1a, df2a, left_index=True, right_index=True)
#or
df1a.join(df2a)

In [ ]:
#enter solution


10.9 Merge the following two df's, getting only the intersection of the two df's

In [ ]:
df6 = pd.DataFrame({"name": ["Peter", "Paul", "Mary"], 
                    "food": ["fish", "beans", "bread"]}, 
                   columns= ["name", "food"])
df7 = pd.DataFrame({"name": ["Mary", "Joseph"], 
                    "drink": ["wine", "beer"]},
                   columns= ["name", "drink"])
print(df6), print(df7), print(pd.merge(df6,df7))

In [ ]:
df6 = pd.DataFrame({"name": ["Peter", "Paul", "Mary"], 
                    "food": ["fish", "beans", "bread"]}, 
                   columns= ["name", "food"])
df7 = pd.DataFrame({"name": ["Mary", "Joseph"], 
                    "drink": ["wine", "beer"]},
                   columns= ["name", "drink"])

#enter solution


10.10 Merge **df6** and **df7**, getting the union of the two df's

In [ ]:
pd.merge(df6,df7, how= "outer")

In [ ]:
#enter solution


10.11 Merge **df6** and **df7**, getting only range of df6

In [ ]:
pd.merge(df6,df7, how="left")

In [ ]:
#enter solution


10.11 Merge **df6** and **df7**, getting only range of df7

In [ ]:
pd.merge(df6,df7, how="right")

In [ ]:
#enter solution


10.12 Merge the following df's, getting both data for the same column name 

In [ ]:
df8 = pd.DataFrame({"name": ["Bob", "Jake", "Lisa", "Sue"], 
                   "rank": [1,2,3,4]})
df9 = pd.DataFrame({"name": ["Bob", "Jake", "Lisa", "Sue"],
                   "rank": [3,1,4,2]})

print(df8), print(df9), print(pd.merge(df8,df9, on="name"))

In [ ]:
df8 = pd.DataFrame({"name": ["Bob", "Jake", "Lisa", "Sue"], 
                   "rank": [1,2,3,4]})
df9 = pd.DataFrame({"name": ["Bob", "Jake", "Lisa", "Sue"],
                   "rank": [3,1,4,2]})

#enter solution


10.13 Merge **df8** and **df9**, getting both data for the same column name and change the suffix to _L and _R 

In [ ]:
pd.merge(df8,df9, on="name", suffixes= ["_L", "_R"])

In [ ]:
#enter solution


#### Task 11 Example: US States Data

Downloading Datasets from Github

In [ ]:
!curl -0 https://raw.githubusercontent.com/jakevdp/data-USstates/master/state-population.csv
!curl -0 https://raw.githubusercontent.com/jakevdp/data-USstates/master/state-areas.csv
!curl -0 https://raw.githubusercontent.com/jakevdp/data-USstates/master/state-abbrevs.csv

11.1 Save the csv files as follow:
     
     state-population as pop
     state-areas as areas
     state-abbrevs as abbrevs

In [ ]:
pop = pd.read_csv("state-population.csv")
areas = pd.read_csv("state-areas.csv")
abbrevs = pd.read_csv("state-abbrevs.csv")

In [ ]:
#enter solution


11.2 Print the first five rows of each dataset

In [ ]:
print(pop.head()), print(areas.head()), print(abbrevs.head())

In [ ]:
#enter solution


11.3 Merge the df's pop and abbrevs, so you get the following columns:
    
    state/region | ages | year | population | state

In [ ]:
merged = pd.merge(pop, abbrevs, how= "outer", left_on= "state/region", right_on = "abbreviation")
merged = merged.drop("abbreviation", 1)
merged.head()

In [ ]:
#enter solution


11.4 Check if there are any Null values

In [ ]:
merged.isnull().any()

In [ ]:
#enter solution


11.5 Show the data with the Null value

In [ ]:
print(merged[merged["population"].isnull()])
print(merged[merged["state"].isnull()])

In [ ]:
#enter solution


11.6 Find out which regions lack the data

In [ ]:
merged.loc[merged["state"].isnull(), "state/region"].unique()

In [ ]:
#enter solution


11.7 Change "state" of "region" "PR" and "USA" to "Puerto Rico" and "United States", after check again for Null values

In [ ]:
merged.loc[merged["state/region"] == "PR", "state"] = "Puerto Rico"
merged.loc[merged["state/region"] == "USA", "state"] = "United States"
merged.isnull().any()

In [ ]:
#enter solution


11.8 Merge the dataset with the df **areas**

In [ ]:
final = pd.merge(merged,areas, on="state", how="left")
final.head()

In [ ]:
#enter solution


11.9 Check for Null values

In [ ]:
final.isnull().any()

In [ ]:
#enter solution


11.10 Check which regions have a Null value in the area column

In [ ]:
final["state"][final["area (sq. mi)"].isnull()].unique()

In [ ]:
#enter solution


11.11 Drop the rows with the Null values

In [ ]:
final.dropna(inplace= True)
final.head()

In [ ]:
#enter solution


11.12 Query all data, where year is equal to 2010 and ages is equal to "total"

In [ ]:
data2010 = final.query("year == 2010 & ages == 'total'")
data2010.head()

In [ ]:
#enter solution


11.13 Set the column "state" as index

In [ ]:
data2010.set_index("state", inplace= True)
data2010.head()

In [ ]:
#enter solution


11.14 Compute the density and save it in a variable

In [ ]:
density = data2010["population"] / data2010["area (sq. mi)"]
density.head()

In [ ]:
#enter solution


11.15 Display the dataset in order, showing the top 5 highest densities

In [ ]:
density.sort_values(ascending= False, inplace= True)
density.head()

In [ ]:
#enter solution


11.16 Display the lowest 5 densities

In [ ]:
density.tail()

In [ ]:
#enter solution


#### Task 12 Aggregation and Grouping
    
    get the planets dataset from seaborn

12.1 Import seaborn

In [ ]:
import seaborn as sns

In [ ]:
#enter solution


12.2 Save the planets dataset in a variable

In [ ]:
planets = sns.load_dataset("planets")

In [ ]:
#enter solution


12.3 Show the shape of the dataset

In [ ]:
planets.shape

In [ ]:
#enter solution


12.4 Show the first five rows of the dataset

In [ ]:
planets.head()

In [ ]:
#enter solution


12.5 Drop the Null values of the data and show the description of the dataset

In [ ]:
planets.dropna().describe()

In [ ]:
#enter solution


12.6 Group the following df by the column "key" and get the sum for each value

In [ ]:
df = pd.DataFrame({"key": ["A", "B", "C", "A", "B", "C"], "data": range(6)},columns= ["key", "data"])
print(df)
df.groupby("key").sum()

In [ ]:
df = pd.DataFrame({"key": ["A", "B", "C", "A", "B", "C"], "data": range(6)},columns= ["key", "data"])
print(df)

#enter solution


12.7 Get the median of the "orbital_period" column for the df **planets** grouped by the column "method"

In [ ]:
planets.groupby("method")["orbital_period"].median()

In [ ]:
#enter solution


12.8 Get the description for the column "year" of the DataFrame grouped by column "method" 

In [ ]:
planets.groupby("method")["year"].describe()

In [ ]:
#enter solution


12.9 Get the minimun, median and max for each other column grouped by the column "key"

In [ ]:
rng = np.random.RandomState(0)
df = pd.DataFrame({"key": ["A", "B", "C", "A", "B", "C"], 
                  "data1": range(6), 
                  "data2": rng.randint(0,10,6)}, 
                  columns = ["key", "data1", "data2"])
print(df)

df.groupby("key").aggregate(["min", np.median, max])

In [ ]:
rng = np.random.RandomState(0)
df = pd.DataFrame({"key": ["A", "B", "C", "A", "B", "C"], 
                  "data1": range(6), 
                  "data2": rng.randint(0,10,6)}, 
                  columns = ["key", "data1", "data2"])
print(df)

#enter solution


12.10 Use the following function to filter the dataset **df** to get the grouped-by "key" data where the standard-deviation is greater than 4

In [ ]:
def filter_func(x):
    return x["data2"].std() > 4

print(df), print(df.groupby("key").std())
print(df.groupby("key").filter(filter_func))

In [ ]:
def filter_func(x):
    return x["data2"].std() > 4

#enter solution


12.11 Transform the dataset, grouping by "key" and transform the data by compute x - x.mean()

In [ ]:
df.groupby("key").transform(lambda x: x - x.mean())

In [ ]:
#enter solution


12.12 Create a new DataFrame where each unique value in the column "key" gets devided by the sum of the same key

In [ ]:
def norm_by_data2(x):
    # x is a DataFrame of group values
    x["data1"] /= x["data2"].sum()
    return x

print(df); print(df.groupby("key").apply(norm_by_data2))

In [ ]:
def norm_by_data2(x):
    # x is a DataFrame of group values
    x["data1"] /= x["data2"].sum()
    return x

#enter solution


#### Task 13 Pivot Tables

    get the titanic dataset from seaborn

13.1 Import seaborn

In [ ]:
import seaborn as sns

In [ ]:
#enter solution


13.2 Save the titanic dataset in a variable

In [ ]:
titanic = sns.load_dataset("titanic")

In [ ]:
#enter solution


13.3 Show the first five rows of the dataset

In [ ]:
titanic.head()

In [ ]:
#enter solution


13.4 Get the mean of survived people grouped by their sex

In [ ]:
titanic.groupby("sex")[["survived"]].mean()

In [ ]:
#enter solution


13.5 Get the mean of the survived grouped by gender and class

In [ ]:
titanic.pivot_table("survived", index= "sex", columns= "class")

In [ ]:
#enter solution


13.6 Get the mean of the survived grouped by gender, age[0,18,80] and class

In [ ]:
age= pd.cut(titanic["age"], [0,18,80])
titanic.pivot_table("survived", index= ["sex", age], columns= "class")

In [ ]:
#enter solution


13.7 Get the mean of the survived grouped by gender, age[0,18,80] and fare[cut in half], class

In [ ]:
fare = pd.qcut(titanic["fare"], 2)
titanic.pivot_table("survived", index= ["sex", age], columns= [fare, "class"])

In [ ]:
#enter solution


Further coming up next:
    
    -Vectorized String Operations
    -Working with Time Series
    -High-Performance Pandas: eval() and query()